src/ipv6/route6_hlist.c

```diff
@@ -0,0 +1,393 @@
+}
+
+static inline bool rt6_match(const struct route6 *rt6, const struct dp_vs_route6_conf *cf)
+{
+    /* Note: Do not use `ipv6_masked_addr_cmp` here for performance consideration 
+     *      here. We ensure the route6 entry is masked when added to route table. */
+    if (ipv6_addr_cmp(&rt6->rt6_dst.addr, &cf->dst.addr) != 0)
+        return false;
+    if (rt6->rt6_dst.plen != cf->dst.plen)
+        return false;
+    if (rt6->rt6_dev && strlen(cf->ifname) != 0) {
+        struct netif_port *dev;
+        dev = netif_port_get_by_name(cf->ifname);
+        if (!dev || dev->id != rt6->rt6_dev->id)
+            return false;
+    }
+
+    /* other fields to be checked? */
+
+    return true;
+}
+
+static struct route6 *__rt6_hlist_get(const struct dp_vs_route6_conf *cf,
+        struct rt6_hlist **phlist)
+{
+    int hashkey;
+    struct rt6_hlist *hlist;
+    struct route6 *rt6;
+
+    list_for_each_entry(hlist, &this_rt6_htable, node) {
+        if (hlist->plen > cf->dst.plen)
+            continue;
+        if (hlist->plen < cf->dst.plen)
+            break;
+        hashkey = rt6_hlist_hashkey(&cf->dst.addr, hlist->plen, hlist->nbuckets);
+        list_for_each_entry(rt6, &hlist->hlist[hashkey], hnode) {
+            if (rt6_match(rt6, cf)) {
+                if (phlist)
+                    *phlist = hlist;
+                return rt6;
+            }
+        }
+    }
+
+    return NULL;
+}
+
+static inline struct route6 *rt6_hlist_get(const struct dp_vs_route6_conf *cf)
+{
+    return __rt6_hlist_get(cf, NULL);
+}
+
+static int rt6_hlist_add_lcore(const struct dp_vs_route6_conf *cf)
+{
+    struct rt6_hlist *hlist = NULL;
+    struct route6 *rt6;
+    int hashkey;
+    bool hlist_exist = false;
+#ifdef DPVS_ROUTE6_DEBUG
+    char buf[64];
+#endif
+
+    if (rt6_hlist_get(cf))
+        return EDPVS_EXIST;
+
+    list_for_each_entry(hlist, &this_rt6_htable, node) {
+        if (hlist->plen <= cf->dst.plen) {
+            if (hlist->plen == cf->dst.plen)
+                hlist_exist = true;
+            break;
+        }
+    }
+
+    if (!hlist_exist) { /* hlist for this prefix not exist, create it! */
+        int i, nbuckets, size;
+        struct rt6_hlist *new_hlist;
+
+        nbuckets = rt6_hlist_buckets(cf->dst.plen);
+        size = sizeof(struct rt6_hlist) + nbuckets * sizeof(struct list_head);
+        new_hlist = rte_zmalloc_socket("rt6_hlist", size, 0, rte_socket_id());
+        if (unlikely(!new_hlist)) {
+            RTE_LOG(ERR, RT6, "[%d] %s: fail to alloc rt6_hlist\n",
+                    rte_lcore_id(), __func__);
+            return EDPVS_NOMEM;
+        }
+
+        new_hlist->plen = cf->dst.plen;
+        new_hlist->nbuckets = nbuckets;
+        new_hlist->nroutes = 0;
+        for (i = 0; i < nbuckets; i++)
+            INIT_LIST_HEAD(&new_hlist->hlist[i]);
+
+        /* add new_hlist to plen-sorted htable */
+        __list_add(&new_hlist->node, hlist->node.prev, &hlist->node);
```

Comment:
will hlist be NULL?

Reply:
I test OK when `this_rt6_htable` is empty. In this case , `hlist->node` is the list head `this_rt6_htable` itself.